### ISSUE TITLE
packed-refs in flake input causes build failure

### ISSUE THREAD

**Issue by benaryorg** [NixOS/nix#8822]
**Describe the bug**

A flake build fails when an input's git repository is cached locally using packed-refs.

**Steps To Reproduce**

1. build a flake
    - in my case it is a flake based NixOS system
1. in the inputs for said flake include a repository which uses git packed-refs, and pull the entire git repository (e.g. `git+https://[…]`, `github:` for instance only pulls the specific revision as an archive¹)
    - packed-refs is the git-mechanism that packs a large amount of refs (such as branches or other) as is the case with the upstream *nixpkgs* repository into a single file `packed-refs` (avoiding a single file for every ref, which is very IO intensive in comparison)
    - now what's required to reproduce this is that your cache for the repository (in my case `/root/.cache/nix/gitv3/…/`) uses packed-refs for the ref that is used by the flake
    - example: `/root/.cache/nix/gitv3/…/refs/` only contains two *empty* directories `heads` and `tags`, however `/root/.cache/nix/gitv3/…/packed-refs` contains the entire list of refs
    - `git pack-refs --all` command can be used to reproduce this manually, *Additional context* has notes on the organic reproduction of this
    - it can also be demonstrated that *git* itself is perfectly fine with this by running `git -C /root/.cache/nix/gitv3/17g5vwq4jjpg5fl4amnvb5cgh6hqx4g4cak8cgyjq6ciji97c9s8/ --git-dir . log ${corresponding_ref?}`; this will succeed, yet the flake build will fail

TL;DR: build a flake using at least one git repository as an input (e.g. `git+https://github.com/NixOS/nixpkgs.git`) and then run `git pack-refs --all` in the cache directory (`~/.cache/nix/gitv3/…/`) for that repository, the flake build will now fail

**Expected behavior**

The flake builds, just as it would with a non-packed ref.

**Actual behaviour**

Full command:

```
nix --extra-experimental-features "nix-command flakes" build --out-link /tmp/nixos-rebuild.dM0XcX/nixos-rebuild "/etc/nixos#nixosConfigurations.\"nixos-builder\".config.system.build.nixos-rebuild" -v --refresh --update-input benaryorg --update-input nixpkgs --commit-lock-file --recreate-lock-file --show-trace
```

Corresponding flake: https://shell.cloud.bsocat.net/infra/blob/e2f82e1b404074d19ec3f4166b84a2d377256814/flake.nix#L-4 (the nixpkgs repository is the one that exhibits the packed-refs here)

```text
warning: could not update mtime for file '/root/.cache/nix/gitv3/17g5vwq4jjpg5fl4amnvb5cgh6hqx4g4cak8cgyjq6ciji97c9s8/refs/heads/nixos-23.05': No such file or directory
error:
       … while updating the lock file of flake 'path:/etc/nixos?lastModified=1691797924&narHash=sha256-XTPAD++VWZ9wGpflk91v2NdV1BY0s9+ZMtLCPNvZc04='

       … while updating the flake input 'nixpkgs'

       … while fetching the input 'git+https://shell.cloud.bsocat.net/nixpkgs?ref=nixos-23.05'

       error: opening file '/root/.cache/nix/gitv3/17g5vwq4jjpg5fl4amnvb5cgh6hqx4g4cak8cgyjq6ciji97c9s8/refs/heads/nixos-23.05': No such file or directory
```

**`nix-env --version` output**

```text
nix-env (Nix) 2.13.5
```

**Additional context**

I honestly don't know why the refs on that one machine of mine are packed. All other machines seem to only have a single or two refs at all (`nixos-23.05` and some still have `nixos-22.05`), so I would assume it's got to do with internal git housekeeping (or gc) or maybe just that I've pulled enough separate other branches on that one machine for git to decide to pack the refs here.
At the very least I don't remember running any manual commands on that cache.
In either case I think nix should be able to handle the packed-ref, ideally by not touching the file directly rather than using the git command (or an appropriate library) to extract the rev for any given ref.

The mtime warning (which is related, but not fatal) happens here:

https://github.com/NixOS/nix/blob/584ff408a4e2b6eb0d74989778c640a92089b6d4/src/libfetchers/git.cc#L573

And the fatal error occurs on this read (which does ultimately fail from utils.cc, but given the error this is the relevant code):

https://github.com/NixOS/nix/blob/584ff408a4e2b6eb0d74989778c640a92089b6d4/src/libfetchers/git.cc#L579

The if around the above code is *not* skipped, as the *strace* output of the fetch shows that the same steps for this repository are followed as with the last up until the point the fatal occurs, and the `rev-parse --is-shallow-repository` is not called for the last repository (strace can be provided on request, same with the repository, however the cached version of this repository is pretty large, it's easier to just use any version of nixpkgs and then run `git -C /root/.cache/nix/gitv3/…/ --git-dir . pack-refs --all` on the cache directory).

It would probably suffice as a temporary workaround to delete the cache and pull the repository anew, but considering this has heavy memory, CPU, network, and IO implications (it *is* nixpkgs after all) I'd like to avoid this.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by bouk** [comment]
I'm experiencing while running the `nixos-upgrade` service on our systems which is annoying. I've created the following (super hacky) workaround for it:


```bash
# Workaround for https://github.com/NixOS/nix/issues/8822

repo=/root/.cache/nix/gitv3/<cache path>

branch=master
head_path="$repo/refs/heads/$branch"
packed_refs_path="$repo/packed-refs"

if [ ! -f "$master_head_path" ] && [ -f "$packed_refs_path" ]; then
    ref=$(grep "refs/heads/$branch" $packed_refs_path | cut -d ' ' -f 1)

    # Create the directory if it doesn't exist
    mkdir -p $repo/refs/heads

    # Write the master ref to refs/heads/$branch
    echo $ref > $head_path
fi
```

I put this script in the `preStart` for nixos-upgrade.service and then it runs fine again.

**Comment by thufschmitt** [comment]
@bouk which version are you experiencing that on? `master` and 2.20 (at least) are now using `libgit` for a bunch of Git-related operations, so it might have been fixed here

**Comment by bouk** [comment]
This is version `nix (Nix) 2.17.1`

Master still seems to be reading directly from the localRefFile: https://github.com/NixOS/nix/blob/61b006ddf644f9a477c4654916069ac0b3786f54/src/libfetchers/git.cc#L588

So I don't think this has been fixed.

**Comment by bouk** [comment]
We keep running into this so I'm going to take a stab at it

**Comment by bouk** [comment]
Fix at https://github.com/NixOS/nix/pull/10241
